### DISCUSSION TITLE
Massive deployment of services causes some of them not to start

### DISCUSSION THREAD

**Discussion by theliuk** [mia-platform/community#277] [Q&A] [answered]
### Which section?

Console

### Which is the version?

v11

### Description

Hi,
in a single project we have `47` distinct micro-services, each having the same image, which is [this](nexus.mia-platform.eu/mia-care/plugins/fhir-api-server), which is a Java service.
Each micro-service has the same CPU and memory requests and limits, being:

**CPU**: 50m req and 1000m limit
**Memory**: 500Mi req and 2000Mi limit

The problem si the following: when deploying each micro-service at once, by deploying all with replicas set to `0` and then using `kubectl scale deployment <micro-service-deploy-name> --replicas=1`, all works out, meaning that all the pods start and stay alive. 

This is not the case whenever I execute a massive deploy, by setting replicas all to `1` and hitting `deploy` in Console. 
If I do so, some of the services start, some of them don't, and some of those who start are eventually restarted because of liveness failure (the service, unfortunately, doesn't log enough to understand the cause of the liveness failure).

Being that all the instances have to connect to the same PostgreSQL cluster, I initially though that the database was the bottleneck. Indeed, it had a maximum number of connection which didn't support all the instances, but then we increased the maximum number of connection and the problem wasn't solved. Other database parameters don't point to the database being the bottleneck.

The cluster where the micro-services are deployed is big enough (it even has autoscaling of nodes enabled). 

I believe this strange behavior has something to do with the throttling that each micro-service suffer at startup, maybe because there's a lot of containers with the same CPU quota.

I wanted to ask if you ever faced the necessity to deploy a massive number of pods on the same cluster, and if you ever experienced an issue of the same kind I'm describing. 

Thank you.

### Actual Outcome

Some pods never really start.

### Expected Outcome

Pods live happily.

**Comment by fredmaggiowski** [comment]
Hi @theliuk, I think the problem is the one you already said, if service startup is CPU intensive they may spike leaving two possibilies:

 - long deployment times since kubernetes will have to wait for spikes to end to find nodes with enough cpu req to schedule new pods
 - liveness probes failure, where cpu is throttled due to limit constraints boot time may take too long and liveness probes start failing (causing the pod being killed and with many pods restarting this may go on forever).

With `kubectl scale deployment <micro-service-deploy-name> --replicas=1` it works because you are scaling one deployment at time, so you won't face this problem (which is mitigated by the latency inherent in the command execution).

What I can suggest you is:

 - increase liveness thresholds to grant the services more time to start
 - increase CPU limits (if you are on a dedicated cluster you might decide to remove them, on a multitenant cluster you must set them)
 - understand whether service boot can be somewhat optimized (if you are using SpringBoot I know it was extremely CPU intensive at boot but I know they made some improvements)
 - pick a different deployment strategy (such as Smart Deploy) and avoid releasing every time 47 pods
 - restructure your project to avoid having that much microservices 😅 

I hope this helps :)

PS in Discussion #263 you can find some tips and considerations about sizing Java resources with Quarkus (if that's what you are using instead)

**Reply by theliuk** [reply]
Thank you @fredmaggiowski. 

Among the solutions if proposed I can apply the 1st and the 2nd.

The service cannot be optimized because is not really ours, it's based on FHIR HAPI. 
Smart deploy is unlikely to be used, since if I change the configuration of one the instance I'll likely have to change others.
I cannot restructure the project because it has to be this way (for the moment).

I will increase liveness probes for sure but I don't know whether removing limits is a good idea, even if I have dedicated cluster.
Isn't it a bad practice to have services without limits?

Thanks, 

Luca

**Reply by fredmaggiowski** [reply]
> Isn't it a bad practice to have services without limits?

The answer, as often is: it depends 😓 There are several opinions on the matter, if you ask for mine it is a bad practice; you should know how your service behaves and avoid the risk of starving other processes for a high request load, you should strive for horizontal and vertical scaling over taking the whole available CPU. 

However there is a lot of literature saying that removing CPU limits is a good thing to make sure you are always available without degrading performances; this is true on paper however it heavily depends on your software architecture and I believe that if your system is not well designed and put under a lot of pressure it will still reach the maximum available CPU and process starts starving. 

So it's really up to you, if you know you don't have that much of traffic, have a dedicated cluster, and the system is well designed you can remove CPU limits and let process take up whatever they need to process their requests.

**Reply by davidetan** [reply]
I agree with everything you wrote.
There's only a point I am not sure it's 100% right:

> long deployment times since kubernetes will have to wait for spikes to end to find nodes with enough cpu req to schedule new pods

Pod scheduling should not care about cpu actual usage on nodes. It should be just a metter of CPU request (and limit since scheduling is also based on QoS, guaranteed comes before burstable that comes before besteffort).

So, for me, if @theliuk excluded external bottlenecks, the problem should be related only to the second possible cause you mentioned:

> liveness probes failure, where cpu is throttled due to limit constraints boot time may take too long and liveness probes start failing (causing the pod being killed and with many pods restarting this may go on forever).

**Reply by davidetan** [reply]
And probably the failure is caused by themselves starving each other.

**Reply by fredmaggiowski** [reply]
>Pod scheduling should not care about cpu actual usage on nodes. 

AFAIK it does, the scheduler reads the CPU request and finds if there is a node with that amount of available CPU. Therefore, Pods may remain Pending due to other processes taking up too much resources.

> It should be just a metter of CPU request (and limit since scheduling is also based on QoS, guaranteed comes before burstable that comes before besteffort).

Yes, however QoS is only for controlling the scheduling priority but with that many services (possibly sharing the same QoS since they are also the same image and apparently heavy coupled in configuration) the long deployment may still happen. Also I assumed QoS was not in the game since @theliuk didn't mention it :)

**Reply by fredmaggiowski** [reply]
>And probably the failure is caused by themselves starving each other.

Agree, I'm pretty sure **this** is the problem

**Reply by davidetan** [reply]
> AFAIK it does, the scheduler reads the CPU request and finds if there is a node with that amount of available CPU. Therefore, Pods may remain Pending due to other processes taking up too much resources.

Maybe the scheduler excludes nodes marked under pressure (since eviction will occur soon), but I'm almost sure that the scheduler just takes into consideration resource requests and allocatable resources on a node.
But I might be wrong. I'll try to do a research lately in the day.

**Reply by fredmaggiowski** [reply]
> But I might be wrong. I'll try to do a research lately in the day.

I might as well, lol; let me know what you find out :)

**Reply by theliuk** [reply]
I believe you all are right. The particular combination of:

1. spikes at startup
2. cluster congestion
3. liveness probes set too early (but I wonder what a good value might be)

creates a cycle where nothing starts. 
The problem is that it is not a linear problem.

I'll consult @daaaaa and then I'll try to figure a way out to solve it at the roots.

Thank you all for your time and attention.

**Reply by fredmaggiowski** [reply]
Great, if you come up with some magic numbers or a finer solution you should definitely join the discussion #263 :)

**Reply by theliuk** [reply]
I met with @daaaaa and this is a summary of our discussion.

The problem happens when a number of services are deployed simultaneously and each replica has CPU consumption spikes at startup (a consumption that differs a lot from the runtime consumption).
CPU request cannot help, because even if you set a very a request, you end up with replicas that will go starving, because all the nodes will support only a subset of all the replicas.

Possibile solutions:

- Accept that each replica might take a lot of time to startup because of the initial throttling, and set set high values for liveness and readiness probes. However, we have to keep in mind that those services that need to connect to external services (like databases) may never connect successfully because of TCP timeouts (falling again in that startup-fail cycle that will never end).

- Spread replicas deployment over time. This is, after all, a synchronization problem, so the best strategy is to spread over time the spikes of CPU consumption.

- Spread replicas deployment over nodes, so that you never have a lot of CPU consumption spikes on the same node. @daaaaa suggested [this kubernetes feature](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/)

**Reply by fredmaggiowski** [reply]
In my opinion, the first and third solution are the ones viable; spread the deployment over time does not solve much and would work against how kubernetes works so I'd not consider it viable to be honest.

**Comment by theliuk** [comment, marked as answer]
@JGiola introduced me to the possibility to use [this property](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes) of k8s deployments which was devised specifically to deal with slow starting containers. 

Basically, with this [startup probe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-startup-probes) we are asking k8s not to bother the container with readiness and liveness probes while it is starting up.

I will try this solution with Kustomize and I'll let you know.

**Reply by theliuk** [reply]
I solved with the usage of `kustomize`, setting the `startupProbe` to a very high value, so that all the pods have the time to startup, even if CPU throttles because of the limits we have to impose.
